<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>
<style>
/* CSS for Markstat 2.0 using Pandoc 2.0 */
body{padding:14px 28px;}
body, table {font-family: Helvetica, Arial, Sans-serif; font-size: 14px;}
h1, h2, h3, h4 {font-weight: normal; color: #3366cc}
h1 {font-size: 200%;}
h2 {font-size: 150%;}
h3 {font-size: 120%;}
h4 {font-size: 100%; font-weight:bold}
img.center {display:block; margin-left:auto; margin-right:auto}
.small{font-size:8pt;}
a {color: black;}
a:visited {color: #808080;}
a.plain {text-decoration:none;}
a.plain:hover {text-decoration:underline;}
.em {font-weight:bold;}
pre, code {font-family: "lucida console", monospace;}
pre.stata {font-size:13px; line-height:13px;}
pre {padding:8px; border:1px solid #c0c0c0; border-radius:8px; background-color:#fdfdfd;}
code {color:#3366cc; background-color:#fafafa;}
pre code { color:black; background-color:white}
/* Added for Pandoc */
figure > img, div.figure > img {display:block; margin:auto}
figcaption, p.caption {text-align:center; font-weight:bold; color:#3366cc;}
h1.title {text-align:center; margin-bottom:0}
p.author, h2.author {font-style:italic; text-align:center;margin-top:4px;margin-bottom:0}
p.date, h3.date {text-align:center;margin-top:4px; margin-bottom:0}
/* Tables*/
table { margin:auto; border-collapse:collapse; }
table caption { margin-bottom:1ex;}
th, td { padding:4px 6px;}
thead tr:first-child th {border-top:1px solid black; padding-top:6px}
thead tr:last-child  th {padding-bottom:6px}
tbody tr:first-child td {border-top:1px solid black; padding-top:6px}
tbody tr:last-child  td {padding-bottom:6px;}
table tbody:last-child tr:last-child td {border-bottom:1px solid black;}
</style>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <title>markstat example</title>
  <style>
      code{white-space: pre-wrap;}
      span.smallcaps{font-variant: small-caps;}
      span.underline{text-decoration: underline;}
      div.column{display: inline-block; vertical-align: top; width: 50%;}
  </style>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
</head>
<body>
<header id="title-block-header">
<h1 class="title">markstat example</h1>
</header>
<h2 id="introduction">Introduction</h2>
<p>This example shows how to run R and Stata code within the same Stata Markdown (<code>.stmd</code>) script. The general approach is detailed on the Stata Markdown website <a href="https://data.princeton.edu/stata/markdown/quantiles">here</a> and <a href="https://data.princeton.edu/stata/markdown/gettingStarted#R">here</a>.</p>
<p>This means that we can use the functions provided by the <a href="https://github.com/MRCIEU/TwoSampleMR"><code>TwoSampleMR</code></a> package to obtain data from <a href="http://www.mrbase.org/">MR-Base</a>.</p>
<p>To see the <code>stmd</code> code which generates this page see <a href="https://raw.githubusercontent.com/remlapmot/mrrobust/master/_drafts/markstat-call-R-example/markstat-call-R-example.stmd">here</a>. You will see that the file is written in markdown and includes R and Stata code chunks.</p>
<p>Before you start please install the following two Stata packages from the SSC archive, so in Stata issue the following commands (I have commented them out because I have already installed them).</p>
<pre class='stata'>. * ssc install whereis
. * ssc install markstat
</pre>
<p>We first need to register the R executable with Stata.</p>
<pre class='stata'>. whereis R "C:\\Program Files\\R\\R-3.6.0\\bin\\R.exe"
C:\\Program Files\\R\\R-3.6.0\\bin\\R.exe
</pre>
<p>Next we have an R code chunk in which we install the required packages in R (again I have commented these lines out because I already have them installed).</p>
<pre class='r'>> # install.packages("devtools")
> # devtools::install_github("MRCIEU/TwoSampleMR")
> # devtools::install_github("MRCIEU/MRInstruments")
</pre>
<p>We will be running the script from the MR-Base paper (<a href="https://doi.org/10.7554/eLife.34408">Hemani et al., 2018</a>). The R code we will use is from <a href="https://raw.githubusercontent.com/explodecomputer/mr-base-methods-paper/master/scripts/ldl-chd.R">here</a>.</p>
<p>First, we load the packages into our R session. Note that the <code>foreign</code> package provides the <code>write.dta()</code> function which we will use to save the data in Stata format.</p>
<pre class='r'>> library(TwoSampleMR)
> library(MRInstruments)
> library(foreign)
</pre>
<p>Our edited version of the code starts by reading in some code to generate a set of plots in R.</p>
<pre class='r'>> source("mrplots.R")
</pre>
<p>We can access the data using the <code>MRInstruments</code> package.</p>
<pre class='r'>> data(gwas_catalog)
> 
> # Get published SNPs for LDL cholesterol
> ldl_snps &lt;- subset(gwas_catalog, grepl("LDL choles", Phenotype) &amp; Author == "Willer CJ")$SNP
> 
> # Extract from GLGC dataset
> exposure &lt;- convert_outcome_to_exposure(extract_outcome_data(ldl_snps, 300))
> 
> # Get outcome data from Cardiogram 2015
> outcome &lt;- extract_outcome_data(exposure$SNP, 7)
> 
> # Harmonise exposure and outcome datasets
> # Assume alleles are on the forward strand
> dat &lt;- harmonise_data(exposure, outcome, action=1)
</pre>
<p>At this point we have our harmonised genotype-exposure and genotype-outcome association data saved in an object in our R session called <code>dat</code>.</p>
<p>The next two code chunks perform the analysis in R.</p>
<pre class='r'>> # Perform MR
> mr(dat)
  id.exposure id.outcome                        outcome
1         300          7 Coronary heart disease || id:7
2         300          7 Coronary heart disease || id:7
3         300          7 Coronary heart disease || id:7
4         300          7 Coronary heart disease || id:7
5         300          7 Coronary heart disease || id:7
                   exposure                    method nsnp         b         se
1 LDL cholesterol || id:300                  MR Egger   62 0.5854136 0.06182590
2 LDL cholesterol || id:300           Weighted median   62 0.4887319 0.03883143
3 LDL cholesterol || id:300 Inverse variance weighted   62 0.4686211 0.03919370
4 LDL cholesterol || id:300               Simple mode   62 0.4678942 0.06358290
5 LDL cholesterol || id:300             Weighted mode   62 0.5189450 0.03371004
          pval
1 1.619410e-13
2 2.521675e-36
3 6.000986e-33
4 5.593001e-10
5 1.109245e-22
> mr_heterogeneity(dat)
  id.exposure id.outcome                        outcome
1         300          7 Coronary heart disease || id:7
2         300          7 Coronary heart disease || id:7
                   exposure                    method        Q Q_df
1 LDL cholesterol || id:300                  MR Egger 170.4804   60
2 LDL cholesterol || id:300 Inverse variance weighted 186.6560   61
        Q_pval
1 1.583556e-12
2 1.154072e-14
> dat$exposure &lt;- "LDL cholesterol"
> dat$outcome &lt;- "Coronary heart disease"
> 
> # Label outliers and create plots
> dat$labels &lt;- dat$SNP
> dat$labels[! dat$SNP %in% c("rs11065987", "rs1250229", "rs4530754")] &lt;- NA
</pre>
<pre class='r'>> png("ldl-chd.png", width=1000, height=1000)
> mr_plots(dat)
> dev.off()
null device 
          1 
</pre>
<figure>
<img src="ldl-chd.png" alt="Plots generated TwoSampleMR." /><figcaption>Plots generated <code>TwoSampleMR</code>.</figcaption>
</figure>
<p>We now save our <code>dat</code> object as a Stata dataset.</p>
<pre class='r'>> write.dta(dat, file = "dat.dta")
</pre>
<p>We now switch from using R code chunks to Stata code chunks. We read the data into Stata and list the variable names (note any <code>.</code> in the colnames of <code>dat</code> have been replaced with <code>_</code>).</p>
<pre class='stata'>. use dat, clear
(Written by R.              )

. ds, v(28)
SNP                     eaf_exposure            samplesize_outcome      year_outcome            data_source_outcome     proxy_a2_outcome        units_exposure_dat
effect_allele_exposure  eaf_outcome             ncase_outcome           pmid_outcome            proxy_outcome           exposure                id_exposure
other_allele_exposure   remove                  ncontrol_outcome        category_outcome        target_snp_outcome      se_exposure             action
effect_allele_outcome   palindromic             pval_outcome            subcategory_outcome     proxy_snp_outcome       pval_exposure           mr_keep
other_allele_outcome    ambiguous               units_outcome           originalname_outcome    target_a1_outcome       units_exposure          labels
beta_exposure           id_outcome              outcome                 outcome_deprecated      target_a2_outcome       mr_keep_exposure
beta_outcome            se_outcome              consortium_outcome      mr_keep_outcome         proxy_a1_outcome        pval_origin_exposure

. di _N
62
</pre>
<p>We can then run the IVW model using <code>mregger</code> with multiplicative standard errors.</p>
<pre class='stata'>. mregger beta_outcome beta_exposure [aw=1/(se_outcome^2)], ivw

                                                      Number of genotypes = 62
──────────────┬────────────────────────────────────────────────────────────────
              │      Coef.   Std. Err.      z    P>|z|     [95% Conf. Interval]
──────────────┼────────────────────────────────────────────────────────────────
beta_outcome  │
beta_exposure │   .4686211   .0391937    11.96   0.000     .3918029    .5454394
──────────────┴────────────────────────────────────────────────────────────────
</pre>
<p>It is helpful to view the forest plot of genotype specific IV estimates.</p>
<pre class='stata'>. decode SNP, gen(rsid)

. mrforest beta_outcome se_outcome beta_exposure se_exposure, ivid(rsid)

. graph export ldl-chd-mrforest.svg, width(600) replace
(file ldl-chd-mrforest.svg written in SVG format)
</pre>
<figure>
<img src="ldl-chd-mrforest.svg" alt="Forest plot of genotype specific IV estimates." /><figcaption>Forest plot of genotype specific IV estimates.</figcaption>
</figure>
<p>We can visualise this model with <code>mreggerplot</code>.</p>
<pre class='stata'>. mreggerplot beta_outcome se_outcome beta_exposure se_exposure

. graph export ldl-chd-mreggerplot.svg, width(600) replace
(file ldl-chd-mreggerplot.svg written in SVG format)
</pre>
<figure>
<img src="ldl-chd-mreggerplot.svg" alt="Plot of the MR-Egger model." /><figcaption>Plot of the MR-Egger model.</figcaption>
</figure>
<p>We then fit the MR-Egger, median, and modal based estimators.</p>
<pre class='stata'>. mregger beta_outcome beta_exposure [aw=1/(se_outcome^2)]

                                                      Number of genotypes = 62
─────────────┬────────────────────────────────────────────────────────────────
             │      Coef.   Std. Err.      z    P>|z|     [95% Conf. Interval]
─────────────┼────────────────────────────────────────────────────────────────
sign(beta~re)│
beta_outcome │
       slope │   .5854136   .0618259     9.47   0.000     .4642371    .7065902
       _cons │  -.0095539   .0040042    -2.39   0.017    -.0174019   -.0017059
─────────────┴────────────────────────────────────────────────────────────────
Residual standard error:  1.686
</pre>
<pre class='stata'>. mrmedian beta_outcome se_outcome beta_exposure se_exposure, weighted

                                                      Number of genotypes = 62
                                                           Replications = 1000
─────────────┬────────────────────────────────────────────────────────────────
             │      Coef.   Std. Err.      z    P>|z|     [95% Conf. Interval]
─────────────┼────────────────────────────────────────────────────────────────
        beta │   .4887683   .0363616    13.44   0.000     .4175009    .5600357
─────────────┴────────────────────────────────────────────────────────────────
</pre>
<pre class='stata'>. mrmodal beta_outcome se_outcome beta_exposure se_exposure, weighted

                                                      Number of genotypes = 62
                                                           Replications = 1000
                                                                       Phi = 1
─────────────┬────────────────────────────────────────────────────────────────
             │      Coef.   Std. Err.      z    P>|z|     [95% Conf. Interval]
─────────────┼────────────────────────────────────────────────────────────────
        beta │    .518945   .0372002    13.95   0.000      .446034     .591856
─────────────┴────────────────────────────────────────────────────────────────
</pre>
<p>And we could continue with additional Stata code (or indeed R code) as we liked.</p>
<p>Note to run this <code>.stmd</code> file in Stata we do so with the following command (specifying additional options as required, see <code>help markstat</code> for more information).</p>
<pre><code>markstat using markstat-call-R-example</code></pre>
<h2 id="references">References</h2>
<ul>
<li>Hemani et al. The MR-Base platform supports systematic causal inference across the human phenome. eLife, 2018;7:e34408 <a href="https://doi.org/10.7554/eLife.34408" class="uri">https://doi.org/10.7554/eLife.34408</a></li>
</ul>
</body>
</html>
